commit 55368156b62c54f2e4e2d18d60c6f58ddb5dbe99
Author: Simon Kaegi <skaegi>
Date:   Tue May 6 17:20:38 2008 +0000

    Bug 222957 IProfileRegistry.add() API insufficient

5	1	bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvCommandProvider.java
1	1	bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvisioningHelper.java
6	5	bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/p2/engine/SimpleProfileRegistry.java
8	7	bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/provisional/p2/engine/IProfileRegistry.java
1	1	bundles/org.eclipse.equinox.p2.installer/src/org/eclipse/equinox/internal/p2/installer/InstallUpdateProductOperation.java
6	1	bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/AbstractProvisioningTest.java
diff --git a/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvCommandProvider.java b/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvCommandProvider.java
index abcb2c4..7826ee3 100644
--- a/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvCommandProvider.java
+++ b/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvCommandProvider.java
@@ -157,7 +157,11 @@ public class ProvCommandProvider implements CommandProvider {
 		if (environments != null)
 			props.setProperty(IProfile.PROP_ENVIRONMENTS, environments);
 
-		ProvisioningHelper.addProfile(profileId, props);
+		try {
+			ProvisioningHelper.addProfile(profileId, props);
+		} catch (ProvisionException e) {
+			interpreter.println("add profile failed " + e.getMessage());
+		}
 	}
 
 	/**
diff --git a/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvisioningHelper.java b/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvisioningHelper.java
index 2b6eba6..6f8379b 100644
--- a/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvisioningHelper.java
+++ b/bundles/org.eclipse.equinox.p2.console/src/org/eclipse/equinox/internal/p2/console/ProvisioningHelper.java
@@ -96,7 +96,7 @@ public class ProvisioningHelper {
 		manager.removeRepository(location);
 	}
 
-	public static IProfile addProfile(String profileId, Properties properties) {
+	public static IProfile addProfile(String profileId, Properties properties) throws ProvisionException {
 		IProfileRegistry profileRegistry = (IProfileRegistry) ServiceHelper.getService(Activator.getContext(), IProfileRegistry.class.getName());
 		if (profileRegistry == null)
 			return null;
diff --git a/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/p2/engine/SimpleProfileRegistry.java b/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/p2/engine/SimpleProfileRegistry.java
index 4e087f0..1a42b26 100644
--- a/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/p2/engine/SimpleProfileRegistry.java
+++ b/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/p2/engine/SimpleProfileRegistry.java
@@ -18,6 +18,7 @@ import javax.xml.parsers.ParserConfigurationException;
 import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Status;
 import org.eclipse.equinox.internal.p2.core.helpers.*;
+import org.eclipse.equinox.internal.provisional.p2.core.ProvisionException;
 import org.eclipse.equinox.internal.provisional.p2.core.eventbus.IProvisioningEventBus;
 import org.eclipse.equinox.internal.provisional.p2.core.location.AgentLocation;
 import org.eclipse.equinox.internal.provisional.p2.engine.*;
@@ -246,20 +247,20 @@ public class SimpleProfileRegistry implements IProfileRegistry {
 		broadcastChangeEvent(id, ProfileEvent.CHANGED);
 	}
 
-	public IProfile addProfile(String id) {
+	public IProfile addProfile(String id) throws ProvisionException {
 		return addProfile(id, null, null);
 	}
 
-	public IProfile addProfile(String id, Map profileProperties) {
+	public IProfile addProfile(String id, Map profileProperties) throws ProvisionException {
 		return addProfile(id, profileProperties, null);
 	}
 
-	public synchronized IProfile addProfile(String id, Map profileProperties, String parentId) {
+	public synchronized IProfile addProfile(String id, Map profileProperties, String parentId) throws ProvisionException {
 		if (SELF.equals(id))
 			id = self;
 		Map profileMap = getProfileMap();
 		if (profileMap.get(id) != null)
-			throw new IllegalArgumentException(NLS.bind(Messages.Profile_Duplicate_Root_Profile_Id, id));
+			throw new ProvisionException(NLS.bind(Messages.Profile_Duplicate_Root_Profile_Id, id));
 
 		Profile parent = null;
 		if (parentId != null) {
@@ -267,7 +268,7 @@ public class SimpleProfileRegistry implements IProfileRegistry {
 				parentId = self;
 			parent = (Profile) profileMap.get(parentId);
 			if (parent == null)
-				throw new IllegalArgumentException(NLS.bind(Messages.Profile_Parent_Not_Found, parentId));
+				throw new ProvisionException(NLS.bind(Messages.Profile_Parent_Not_Found, parentId));
 		}
 
 		Profile profile = new Profile(id, parent, profileProperties);
diff --git a/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/provisional/p2/engine/IProfileRegistry.java b/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/provisional/p2/engine/IProfileRegistry.java
index bdf3946..2a03ffe 100644
--- a/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/provisional/p2/engine/IProfileRegistry.java
+++ b/bundles/org.eclipse.equinox.p2.engine/src/org/eclipse/equinox/internal/provisional/p2/engine/IProfileRegistry.java
@@ -11,6 +11,7 @@
 package org.eclipse.equinox.internal.provisional.p2.engine;
 
 import java.util.Map;
+import org.eclipse.equinox.internal.provisional.p2.core.ProvisionException;
 
 /**
  * This encapsulates the access to the profile registry. 
@@ -41,10 +42,10 @@ public interface IProfileRegistry {
 	 * 
 	 * @param id the profile id
 	 * 
-	 * @throws IllegalArgumentException if a profile
+	 * @throws ProvisionException if a profile
 	 *         with the same id is already present in the registry.
 	 */
-	IProfile addProfile(String id);
+	IProfile addProfile(String id) throws ProvisionException;
 
 	/**
 	 * Add the given profile to this profile registry.
@@ -52,10 +53,10 @@ public interface IProfileRegistry {
 	 * @param id the profile id
 	 * @param properties the profile properties
 	 * 
-	 * @throws IllegalArgumentException if a profile
+	 * @throws ProvisionException if a profile
 	 *         with the same id is already present in the registry.
 	 */
-	IProfile addProfile(String id, Map properties);
+	IProfile addProfile(String id, Map properties) throws ProvisionException;
 
 	/**
 	 * Add the given profile to this profile registry.
@@ -64,10 +65,10 @@ public interface IProfileRegistry {
 	 * @param properties the profile properties
 	 * @param parentId the id of a parent profile
 	 * 
-	 * @throws IllegalArgumentException if a profile
-	 *         with the same id is already present in the registry.
+	 * @throws ProvisionException if a profile
+	 *         with the same id is already present in the registry or the parentId is not a registered Profile.
 	 */
-	IProfile addProfile(String id, Map properties, String parentId);
+	IProfile addProfile(String id, Map properties, String parentId) throws ProvisionException;
 
 	/**
 	 * Remove the given profile from this profile registry.
diff --git a/bundles/org.eclipse.equinox.p2.installer/src/org/eclipse/equinox/internal/p2/installer/InstallUpdateProductOperation.java b/bundles/org.eclipse.equinox.p2.installer/src/org/eclipse/equinox/internal/p2/installer/InstallUpdateProductOperation.java
index 772a218..26012e6 100644
--- a/bundles/org.eclipse.equinox.p2.installer/src/org/eclipse/equinox/internal/p2/installer/InstallUpdateProductOperation.java
+++ b/bundles/org.eclipse.equinox.p2.installer/src/org/eclipse/equinox/internal/p2/installer/InstallUpdateProductOperation.java
@@ -70,7 +70,7 @@ public class InstallUpdateProductOperation implements IInstallOperation {
 	/**
 	 * Create and return the profile into which units will be installed.
 	 */
-	private IProfile createProfile() {
+	private IProfile createProfile() throws ProvisionException {
 		IProfile profile = getProfile();
 		if (profile == null) {
 			Map properties = new HashMap();
diff --git a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/AbstractProvisioningTest.java b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/AbstractProvisioningTest.java
index 5bcbffc..080e068 100644
--- a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/AbstractProvisioningTest.java
+++ b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/AbstractProvisioningTest.java
@@ -15,6 +15,7 @@ import junit.framework.TestCase;
 import org.eclipse.core.runtime.*;
 import org.eclipse.equinox.internal.p2.core.helpers.ServiceHelper;
 import org.eclipse.equinox.internal.p2.metadata.repository.MetadataRepositoryManager;
+import org.eclipse.equinox.internal.provisional.p2.core.ProvisionException;
 import org.eclipse.equinox.internal.provisional.p2.director.*;
 import org.eclipse.equinox.internal.provisional.p2.engine.*;
 import org.eclipse.equinox.internal.provisional.p2.metadata.*;
@@ -593,7 +594,11 @@ public abstract class AbstractProvisioningTest extends TestCase {
 		IProfileRegistry profileRegistry = (IProfileRegistry) ServiceHelper.getService(TestActivator.getContext(), IProfileRegistry.class.getName());
 		profileRegistry.removeProfile(name);
 		//create and return a new profile
-		return profileRegistry.addProfile(name, properties, parentId);
+		try {
+			return profileRegistry.addProfile(name, properties, parentId);
+		} catch (ProvisionException e) {
+			throw new IllegalArgumentException(e.getMessage());
+		}
 	}
 
 	protected IProgressMonitor getMonitor() {
